### ISSUE TITLE
Using the monodepth2 output as input to 3D object detection for the kitti object dataset

### ISSUE THREAD

**Issue by tareeqav** [nianticlabs/monodepth2#96]
I apologize if this is outside the scope of what you're willing to discuss; which is entirely understandable.

I successfully generated lidar point clouds using the output from your `mono+stereo_1024x320` model and the [code from](https://github.com/mileyan/pseudo_lidar/blob/master/preprocessing/generate_lidar.py#L10) from the pseudo-lidar paper.

As you can see in the link referenced, they define two methods:

```
def project_disp_to_points:

# this is meant to be used from output of the DORN network
# https://github.com/hufu6371/DORN/
def project_depth_to_points: 
```

Therefore I have a question about how to correctly use the output of your network to do this step. I have attempted the following:

- using depth from predicted disparity 
```(python)
disp = outputs[("disp", 0)]
disp_resized = torch.nn.functional.interpolate(
    disp, (original_height, original_width), mode="bilinear", align_corners=False)

pred_disp, _ = disp_to_depth(disp_resized, 0.1, 100)
pred_disp = pred_disp.cpu()[:, 0].numpy()

pred_depth = 1 / pred_disp
pred_depth *= 5.4 # TO-DO: remove hard coded value

# then I call the project_depth_to_points with pred_depth
```

- using the scaled predicted disparity
```
disp = outputs[("disp", 0)]
disp_resized = torch.nn.functional.interpolate(
        disp, (original_height, original_width), mode="bilinear", align_corners=False)
scaled_disp, _ = disp_to_depth(disp_resized, 0.1, 100)

# then I call project_disp_to_points using the scaled_disp
```

The latter method produces a more accurate "looking" point cloud, when viewed using `pcl_viewer`.

However neither method produces a good enough result that can be used as input to a 3D object detection from the [PointNet](https://github.com/charlesq34/frustum-pointnets) model.

When I use the ground-truth LiDAR as input to that model, I am getting correctly labeled 3D bounding boxes, however with LiDAR generated above, the 3D bounding boxes are very inaccurate (as compared with the ground truth labels from kitti object dataset).

**Side note**: I have yet to attempt using DORN.  The PSMNet model defined in the `pseudo-lidar` paper is simply too heavy.  I am hoping to be able to use `monodepth2` since it's self-supervised and thus can be trained using images collected by any stereo camera.

Finally if you can comment on the validity of using your model as a substitute to the DORN / PSMNET would be highly appreciated.  Might save me a couple of weeks down the wrong track.

**Comment by tareeqav** [comment]
Just found a [related issue](https://github.com/mrharicot/monodepth/issues/130) from your original paper.

However unlike the previous code which has an unbounded maximum disparity, the current code has one at `100`, correct?

**Comment by mrharicot** [comment]
Hi,

You indeed need to use `scaled_disp`.
`disp` is simply the output of a sigmoid (so bound between 0 and 1) and the `disp_to_depth` [function](https://github.com/nianticlabs/monodepth2/blob/master/layers.py#L16) is used to scale it between two values such that the depth computed from it is bound between `min_depth` (0.1) and `max_depth` (100).
These depth values differ from meters from a scale, hence they are not metric.

Now if you want to get a 3D pointcloud from the depth you need to use the `BackprojectDepth` [class](https://github.com/nianticlabs/monodepth2/blob/master/layers.py#L139) which we use in the [trainer](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L377).

Now real question is: what is your goal? What do you want to do with the pointcloud?

PSMNet is a _stereo_ network and requires two input images, but is more accurate.
DORN is a _mono_ network and, since it was trained with full supervision, it is typically more precise than monodepth2 but suffers from quantization artefacts.

**Comment by tareeqav** [comment]
Hi @mrharicot, thanks so much for the explanation!

My true goal is to replicate the work done [here](https://arxiv.org/pdf/1903.09847.pdf) and [here](https://arxiv.org/pdf/1812.07179.pdf), which is monocular 3D object detection by generating "pseudo-LiDAR".

Both these works "lift" the disparity map to a bird's eye view type point cloud and then use that as input to well established 3D detection networks, such as [Frustom PointNet](https://arxiv.org/pdf/1711.08488.pdf).  

There seems to be a big shift in autonomous vehicles to camera only solutions.  Baidu released [Apollo-Lite](https://www.cnet.com/roadshow/news/baidu-apollo-lite-camera-based-self-driving/) which is a vision only level 4 system, MobileEye and Tesla have argued for the same for some years.  

In the end, all I am doing is learning how to become an autonomous vehicle engineer by self learning and hopefully putting together a really good PoC with a Go-Kart.

**Side Note**:  The Average Precision for pseudo LiDAR methods on Kitti 3D object benchmark is 42% ... far below the ca. 80% of direct LiDAR but arguably very useful in short range.

**Comment by tareeqav** [comment]
Many thanks @mrharicot .  I was able to get reasonable results, but I've thought of another way and thus another question.  Closing issue and hopefully will get a response from you guys on new question :)

**Comment by CaptainEven** [comment]
@tareeqav Your problem is probably due to relative pose transformation(coordinate system transformation). As most lidar point cloud based 3D object detector is trained on lidar coordinate system rather than camera coordinate system, the pseudo lidar point cloud you get need to do the transformation as well. As a result, you may get a "good looking" pseudo lidar point cloud, but not 3D bounding box got! 
Looking forward to more discussion!
